### PR TITLE
feat: add MCP support to plugins

### PR DESCRIPTION
**Pull request by saoudrizwan**
This PR adds MCP as a first-class plugin capability.

Plugins can already bundle Cline primitives such as tools, rules, hooks, skills, and providers. This adds MCP servers to that plugin surface while keeping lifecycle ownership explicit: a plugin can declare an MCP server, Cline can install it into the normal MCP settings model, and Cline can later disable or remove the MCP entry because it knows which plugin owns it.

The current implementation intentionally uses the existing `cline_mcp_settings.json` path instead of a parallel plugin-only MCP runtime. That keeps OAuth, settings loading, MCP UI, enable/disable state, and server startup behavior on the same code path Cline already uses for user-configured MCP servers.

Google Chrome 2026-06-12 16 41 32
Google Chrome 2026-06-12 16 22 34

## User behavior

- Installing or enabling a plugin syncs its declared MCP servers into `cline_mcp_settings.json` under `mcpServers`.
- Plugin-created MCP entries are stamped with metadata: `source: "plugin"`, `pluginName`, and `pluginPath`.
- If a plugin declares an MCP server name that already exists and is not owned by that same plugin, Cline skips it instead of overwriting the user’s server.
- Disabling a plugin marks its owned MCP entries as `disabled: true`.
- Re-enabling the plugin re-syncs those entries and clears the disabled state.
- Uninstalling the plugin removes MCP entries owned by that plugin.
- If a plugin changes or removes an MCP declaration, sync updates or removes the stale owned entry.
- Existing OAuth state on a plugin-owned MCP entry is preserved when the plugin updates the server definition.

## OAuth

OAuth now works through the existing MCP settings-backed flow because plugin MCP servers are regular MCP settings entries with plugin ownership metadata. There is no separate plugin OAuth sidecar and no plugin-owned OAuth store in this PR.

That means a remote plugin MCP server can use the same persisted OAuth state shape Cline already supports in `cline_mcp_settings.json`. Updating or re-syncing the plugin MCP definition preserves the existing `oauth` object, while uninstall removes the full plugin-owned MCP entry.

## Technical approach

Plugins still declare MCP servers with the existing plugin API:

```ts
const plugin = {
  name: "repo-docs",
  manifest: { capabilities: ["mcp"] },
  setup(api) {
    api.registerMcpServer({
      name: "repo-docs",
      transport: {
        type: "streamableHttp",
        url: "https://example.com/mcp",
      },
    })
  },
}
```

The new settings sync layer loads sandboxed plugins, collects `registerMcpServer()` declarations, normalizes them through the existing plugin MCP registration validation, then writes only plugin-owned MCP entries into the normal MCP settings file.

The sync layer is intentionally conservative:

- It does not overwrite user-owned MCP entries on name collision.
- It preserves existing OAuth state when refreshing plugin-owned entries.
- It does not create an MCP settings file for plugins with no MCP servers.
- It does not overwrite invalid MCP settings JSON.
- It only removes entries that are marked as plugin-owned and match the plugin path/name being synced or uninstalled.

## What changed from the original runtime-only direction

The earlier version of this PR treated plugin MCP as an ephemeral runtime contribution and avoided writing user MCP settings. That made ownership clean, but it made OAuth awkward because the current OAuth implementation persists through MCP settings.

This version chooses the simpler path: plugin MCP entries live in the normal settings file, with metadata identifying plugin ownership. That gives us OAuth support now, keeps the MCP manager behavior consistent, and still lets plugin disable/uninstall clean up the owned entries.

## Testing

Verified with focused core and CLI tests covering:

- syncing plugin MCP entries into settings
- preserving OAuth state on update
- skipping user-owned name collisions
- stale plugin MCP removal
- plugin disable/re-enable behavior
- plugin uninstall cleanup
- MCP settings UI loading plugin-owned entries from settings
- removal of direct runtime plugin MCP tool injection

Commands run:

```bash
bun -F @cline/core test:unit -- src/services/plugin-mcp-settings.test.ts src/services/plugin-uninstall.test.ts src/runtime/orchestration/session-runtime-orchestrator.test.ts
bun -F @cline/cli test:unit -- src/runtime/interactive/config-data.test.ts src/commands/plugin.test.ts src/tui/components/dialogs/mcp-manager-dialog.test.ts
bun -F @cline/core typecheck
bun -F @cline/cli typecheck
bun biome check apps/cli/src/commands/plugin.ts apps/cli/src/runtime/interactive/config-data.ts apps/cli/src/runtime/interactive/config-data.test.ts apps/cli/src/tui/hooks/use-mcp-manager.tsx apps/cli/src/tui/interactive-config.ts sdk/packages/core/src/index.ts sdk/packages/core/src/runtime/orchestration/session-runtime-orchestrator.ts sdk/packages/core/src/runtime/orchestration/session-runtime-orchestrator.test.ts sdk/packages/core/src/services/plugin-tools.ts sdk/packages/core/src/services/plugin-uninstall.ts sdk/packages/core/src/services/plugin-mcp-settings.ts sdk/packages/core/src/services/plugin-mcp-settings.test.ts
```

### CODE REVIEW

File: apps/cli/src/runtime/interactive/config-data.ts [hunk 1]
```diff
@@ -71,6 +73,17 @@ export function createInteractiveConfigDataLoader(input: {
...
 			setDisabledPlugin(item.path, item.enabled);
+			if (item.enabled) {
+				disablePluginMcpServersInSettings({ pluginPaths: [item.path] });
+			} else {
+				await syncPluginMcpServersToSettings({
+					pluginPaths: [item.path],
+					cwd: input.config.cwd,
+					workspacePath: workspaceRoot(),
+					providerId: input.config.providerId,
+					modelId: input.config.modelId,
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply apps/cli/src/runtime/interactive/config-data.ts:87]: Addressed in `7b9830763`.

On plugin re-enable, the caller now inspects the `syncPluginMcpServersToSettings()` result when there are existing plugin-owned MCP entries to reactivate. If sync reports failures, we throw with the failure details so the existing TUI toggle path surfaces an inline error instead of silently leaving the MCP server disabled while the plugin appears enabled.

I scoped this to plugins that already own MCP entries so we do not change the existing behavior for unrelated/broken non-MCP plugins.

File: sdk/packages/core/src/services/plugin-mcp-settings.ts [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+					? metadata.pluginName
+					: "plugin",
+			pluginPath:
+				typeof metadata?.pluginPath === "string" ? metadata.pluginPath : "",
+			action: "disabled",
+		});
+	}
+	if (mutations.length > 0) {
+		writeRawSettings(settingsPath, settings, servers);
+	}
```
[the remark this answers is not in the document]
Comment (saoudrizwan) [reply sdk/packages/core/src/services/plugin-mcp-settings.ts:506]: Addressed in `cc7470cfb`.

I kept the low-level MCP settings helpers throwing on write failure and changed the callers to avoid split state instead:

- plugin disable now writes the plugin-owned MCP `disabled: true` state before marking the plugin disabled in global settings
- plugin re-enable now re-syncs MCP before marking the plugin enabled globally
- plugin uninstall now removes plugin-owned MCP entries before deleting plugin files

Added regressions for both failure modes:

- failed MCP disable write does not mark the plugin disabled globally
- failed MCP cleanup write does not remove plugin files

Validation run:

- `bun run --cwd apps/cli test:unit -- src/commands/plugin.test.ts src/runtime/interactive/config-data.test.ts`
- `bun run --cwd sdk/packages/core test:unit -- src/services/plugin-uninstall.test.ts src/services/plugin-mcp-settings.test.ts src/runtime/orchestration/session-runtime-orchestrator.test.ts`
- `bun run --cwd apps/cli typecheck`
- `bun run --cwd sdk/packages/core typecheck`
- manual plugin MCP lifecycle smoke passed
